### DISCUSSION TITLE
Traffic Light Negotiation and Perception-Based Detection (Project 223)

### DISCUSSION THREAD

**Discussion by robertogl** [mathworks/MATLAB-Simulink-Challenge-Project-Hub#55] [Collaborate]
Contribute to the discussion by asking and/or answering questions, commenting, or sharing your ideas for solutions to [project 223](https://github.com/mathworks/MathWorks-Excellence-in-Innovation/tree/main/projects/Traffic%20Light%20Negotiation%20and%20Perception-Based%20Detection)

---
*:robot: An AI bot may reply to your comments with suggestions and guidance. It will not solve the project for you. Your comments may be processed by a third-party LLM. To direct a question to a specific person, tag them with @username.*

**Comment by sai303030** [comment]
i need  some idea about project and abstract of the project

**Reply by robertogl** [reply]
Hi @sai303030,

What do you mean precisely?

**Reply by sai303030** [reply]
I was take up a project under the domain of machine learning and tittle of my project is traffic light negotiation and perception based detection , I need some tips and abstract of the project..

**Comment by nikhiltkumar** [comment]
Where can I find hints for the solution?

**Reply by robertogl** [reply]
Hi @nikhiltkumar,

In the background material section we have links to several examples that can be very useful for starting the project and work towards its completion. Please let us know if there is anything in particular you need help with.

**Comment by Vikas-ABD** [comment]
can I get  any reading papers for this project

**Comment by dikokob** [comment]
Hi @robertogl,

I hope you are well.

I am from the University of Cape Town, and one of my students is doing the project. 

However, it seems we need to purchase Road Runner Scenario. Please let me know if this is included in our licence.

**Reply by robertogl** [reply]
Hello @dikokob,

University of Cape Town should have a full-suite Wide-Campus license, which includes all the products. However, let me just confirm with the account manager.

**Reply by robertogl** [reply]
I got the confirmation that you have Road Runner and Road Runner Scenario as you can see in this [list](https://www.mathworks.com/content/dam/mathworks/fact-sheet/campus-wide-license-products.pdf).  However, Road Runner installation can be a bit tedious. Here the [instructions](https://www.mathworks.com/content/dam/mathworks/mathworks-dot-com/images/responsive/supporting/solutions/automated-driving/roadrunner-tutorial/access-roadrunner-cwl.pdf).

I hope this helps.

**Reply by dikokob** [reply]
@robertogl thanks. I'll let you know how it goes.

I have asked my student to put a bit of a twist on the project by making a South African (i.e. load shedding) scenario.  Would this be eligible for them to submit it for MathWorks Excellence in Innovation?

Reach out on boitumelo.dikoko@uct.ac.za

**Reply by robertogl** [reply]
Of course! That would even introduce extra complexity to the project.

**Comment by megwilson8** [comment]
Hi @robertogl,
I am currently doing my thesis at UCT and the topic is Traffic light negotiation with perception based detection. I think you were communicating with my supervisors above. I am trying to integrate a road runner scenario into the prebuilt traffic light negotiation with unreal engine to create a South African scenario and adapt the Simulink for these conditions. 

I have created a road runner scene and exported it to unreal engine. When I export the RoadRunner scene there are some errors which are effecting the driving scenario. Is there a way to fix these errors so I can implement the Roadrunner scene in the prebuilt traffic light negotiation with unreal engine? Is there also a way to find the errors before exporting the file, debugging? 

I have attached the error messages to this message. 

![DrivingScenario](https://github.com/mathworks/MathWorks-Excellence-in-Innovation/assets/88933163/4636c6d6-055d-434d-9dee-fad169d3eadb)
![DrivingScenariomap](https://github.com/mathworks/MathWorks-Excellence-in-Innovation/assets/88933163/3f6f2c2e-d07d-410f-aeec-8d492bba5cc5)
![RoadRunnererror](https://github.com/mathworks/MathWorks-Excellence-in-Innovation/assets/88933163/724d0681-71e8-4611-998b-24b9f931867d)
![RoadRunnererror2](https://github.com/mathworks/MathWorks-Excellence-in-Innovation/assets/88933163/e197906d-a728-4430-95af-c06d3c36c4ce)
![Unreal](https://github.com/mathworks/MathWorks-Excellence-in-Innovation/assets/88933163/233302fb-bf34-45c0-b128-7e5fad2d4f94)

**Reply by pfryscak** [reply]
Hi. Just a quick check. Are you exporting the roads as left-hand drive? https://www.mathworks.com/help/roadrunner/ug/left-hand-drive-export-to-opendrive-1-4.html

**Reply by megwilson8** [reply]
I wasn't at first but now I've exported the Road Runner as left-hand drive and am still getting some errors which come up in the driving Scenario.

![DrivingS](https://github.com/mathworks/MathWorks-Excellence-in-Innovation/assets/88933163/517bcbb5-d1d0-436c-bdc0-4997ab7e26ae)
![New Driving](https://github.com/mathworks/MathWorks-Excellence-in-Innovation/assets/88933163/795fcffb-67ea-4636-96be-8d53f4ca89a2)
![RoadRun](https://github.com/mathworks/MathWorks-Excellence-in-Innovation/assets/88933163/532a8d04-6ce8-4f13-8b80-42e02108659a)
sceario.

**Reply by pfryscak** [reply]
@megwilson8 would you be williing to send us an .rrscene file of your scene? In the meanwhile, kindly click on the link for "Road 40" and "Road 11" in the Output window. For the first, select the "Maneuver Tool" in RoadRunner and for the second, select the "Lane Tool". I'm not sure about the "Road 40" problem, but I suspect that the "Road 11" problem is a lane headed in the wrong direction. Feel free to send screenshots of how those areas look in those tools.

**Reply by megwilson8** [reply]
Thanks for the help with Road 11 that seems to be fixed. Here is the link to my RoadRunner Scene https://github.com/megwilson8/Traffic-Light-Negotiation-.git. Seems to still be giving me many errors when I open it in MATLAB Driving Scenario.

**Reply by pfryscak** [reply]
Thank you. Quick question: what version of MATLAB are you using?

**Reply by megwilson8** [reply]
I am using the 2023a for both RoadRunner and Matlab.

**Reply by pfryscak** [reply]
[Engen Intersection Corner Fixed.zip](https://github.com/mathworks/MathWorks-Excellence-in-Innovation/files/12490481/Engen.Intersection.Corner.Fixed.zip)
Hi! Import into ADT still has issues; ADT doesn't support lanes that taper, for example. I sent an internal message to that team asking what they suggest.

In the meanwhile, attached is a corrected version of your file, that will export from RoadRunner without error. There was an odd issue at one corner near the petrol station that I edited.

**Reply by megwilson8** [reply]
@pfryscak really appreciate the help.
Many thanks!

**Reply by megwilson8** [reply]
@pfryscak I have updated the Road Runner file but still receiving errors when I upload the file into MATLAB Driving Scenario. The main error is the lane width which I'm not sure how to check and edit? I have uploaded both the RoadRunner and Driving Scenario file to github (https://github.com/megwilson8/Traffic-Light-Negotiation-.git )  and attached screenshots of the errors below. 

![Driving Scenario Sunday](https://github.com/mathworks/MATLAB-Simulink-Challenge-Project-Hub/assets/88933163/3616660e-4bbf-4c87-bf2c-36cc4df90323)
![DrivingS error](https://github.com/mathworks/MATLAB-Simulink-Challenge-Project-Hub/assets/88933163/d359be12-ee52-419c-a277-b0ed6d30bfe0)

**Comment by pfryscak** [comment]
Hi. I have not yet received detailed instructions from the Automated Driving Toolbox (ADT) team. In the meanwhile, please take a look at https://www.mathworks.com/help/roadrunner/ref/lanewidthtool.html. Make a copy of your RRSCENE and adjust the lanes to have uniform width along their length and no tapers (forming lanes). Try importing that into ADT and see if it reduces the errors.

**Reply by megwilson8** [reply]
I have put this on hold for the moment and have move to perception based detection of the traffic light states. Using the simulation environment of Traffic Light Negotiation using Unreal Engine I have saved the simulation video and have attempted to train the detector to detect the traffic light states. I used Video Labeler in MATLAB but the detection is very inaccurate sometimes never detecting the traffic light or detecting everything but the traffic light. 
I went with the approach of labeling each state of the traffic light and using this to train the detector but this was very inaccurate. I then tried to train the detection from only one labeled file for instance the video only had labels of green. This had a slightly better response but still very far from anything I can work with. 
Is there a better way to use perception based detection to detect the traffic lights as I am trying to modify the model to be able to detect the traffic light state without using the state flow input.

**Reply by megwilson8** [reply]
@pfryscak

**Reply by pfryscak** [reply]
Hi @megwilson8. Once things are outside of the RoadRunner realm, I'm afraid I have less information. I will see if a colleague can help.

**Reply by robertogl** [reply]
Hi @megwilson8,

There are several factors to consider here. For example, the data that you use to train the detector. You need to have enough data and from different perspective, distance, and background with a good balance for all three lights. Also, what kind of detector are you using and how are you training it? Training a Yolov2 with 3 classes (one per color) should work. Otherwise there are other techniques that do not adopt NN that still work fine.

**Reply by megwilson8** [reply]
Hi @robertogl,
I will try the Yolov2 training. 
I'm not sure how familiar you are with the Traffic Light Negotiation with Unreal Engine Visualization model but I am trying to train a detector to use perception based detection to determine the state of the traffic light. I also need to train the light to detect a broken traffic light (no colour) but I cant find where I could implement this into the model, displaying no colour. 
Do you know how I can implement a scenario where there is  no state showing on the traffic light?

**Comment by zy18879589365** [comment]
Is there a publicly available dataset for training?
